aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +22,333 @@ namespace at {
+    TH_TENSOR_APPLY_hasFinished = true;
+  while (!TH_TENSOR_APPLY_hasFinished) {
     /* Loop through the inner most region of the Tensor */
-    for(; tensor1_i < tensor1_size && tensor2_i < tensor2_size; tensor1_i++, tensor2_i++, tensor1_data += tensor1_stride, tensor2_data += tensor2_stride)
-    {
-      op(*tensor1_data, *tensor2_data);
+    for (; ut1.i < ut1.size && ut2.i < ut2.size;
+         ut1.i++, ut2.i++, ut1.data += ut1.stride, ut2.data += ut2.stride) {
+      op(*ut1.data, *ut2.data);
     }
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor1, 0)
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor2, 0)
+    TH_TENSOR_APPLY_hasFinished = tensor_update_counters(ut1);
+    TH_TENSOR_APPLY_hasFinished = tensor_update_counters(ut2);
   }
-  if(tensor1_counter != NULL)
-    delete [] tensor1_counter;
-  if(tensor2_counter != NULL)
-    delete [] tensor2_counter;
 }
 
 /*
   Apply a pointwise operator to two tensors.
 
-  The calling convention for op is a function/functor that takes takes two references to
-  type scalar; at least one of these references should be non-const in order to write the output.
-  For example, to compute a = b^2, op would be of the form:
+  The calling convention for op is a function/functor that takes takes two
+  references to type scalar; at least one of these references should be
+  non-const in order to write the output. For example, to compute a = b^2, op
+  would be of the form:
   [](scalar &a_val, const scalar &b_val) { a_val = b_val * b_val; };
 */
-template<typename scalar1, typename scalar2, typename Op>
+template <typename scalar1, typename scalar2, typename Op>
 void CPU_tensor_apply2(Tensor tensor1, Tensor tensor2, Op op) {
   CPU_tensor_apply2_dim<scalar1, scalar2, Op>(tensor1, tensor2, -1, op);
 }
 
-template<typename scalar1, typename scalar2, typename scalar3, typename Op>
-void CPU_tensor_apply3_dim(Tensor &tensor1, Tensor& tensor2, Tensor& tensor3, int64_t dim, Op op) {
+template <typename scalar1, typename scalar2, typename scalar3, typename Op>
+void CPU_tensor_apply3_dim(
+    Tensor& tensor1,
+    Tensor& tensor2,
+    Tensor& tensor3,
+    int64_t dim,
+    Op op) {
   checkBackend("CPU_tensor_apply3", {tensor1, tensor2, tensor3}, Backend::CPU);
-  bool TH_TENSOR_APPLY_hasFinished = false;
-  int64_t TH_TENSOR_dim_index = 0;
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar1, tensor1, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar2, tensor2, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar3, tensor3, dim, 1)
+  util_tensor<scalar1> ut1(tensor1, dim, true);
+  util_tensor<scalar2> ut2(tensor2, dim, true);
+  util_tensor<scalar2> ut3(tensor3, dim, true);
 
   int elements_equal = 1;
   auto t1_numel = tensor1.numel();
   auto t2_numel = tensor2.numel();
   auto t3_numel = tensor3.numel();
-  if(t1_numel!= t2_numel) {
+  if (t1_numel != t2_numel) {
     elements_equal = 0;
-  } else if(t1_numel != t3_numel) {
+  } else if (t1_numel != t3_numel) {
     elements_equal = 0;
   }
   if (elements_equal == 0) {
```

Comment:
Do we really need those `if`s above? We could just do `if (t1_numel != t2_numel || t1_numel != t3_numel) { // error }`. This pattern could be improved in all the apply functions